### PR TITLE
Android Automation setup docs - for review

### PR DESCRIPTION
**Pull request by mitchhentges**
Don't merge this PR - the "actual" PR-ing of this documentation will occur after the [docs RFC](https://github.com/mozilla-releng/releng-rfcs/pull/7) is resolved.

[Rendered](https://github.com/mitchhentges/build-relengdocs/blob/dont-merge-automation-instructions/procedures/automation-android-setup.md)

### CODE REVIEW

File: procedures/automation-android-setup.md [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+Note: we don't need to explicitly define scopes in Taskcluster. We just define them in the task, hook's role, and each scriptworker script. Taskcluster will implicitly detect scopes, but it doesn't use them for anything other than generic equality/wildcard verification.
+
+1. Request dep and release signing keys (example bug for [`reference-browser`](https://bugzilla.mozilla.org/show_bug.cgi?id=1508761)). Confirm with the app's team what the "signature algorith" and "digest algorithm" should be, and include that information in the ticket.
+2. Clone the product's repository
+    * Note that, while doing "staging" runs, you'll have to use your `master` branch in your fork. Hook cloning/CoT/etc haven't been tested with a non-default branch
+3. Add `.taskcluster.yml` in the root of the repository if it doesn't exist, and add a task for cron jobs. ([`fenix` example, go to `tasks_for == "cron"`](https://github.com/mozilla-mobile/fenix/blob/master/.taskcluster.yml))
+    1. Update the `event.repository.clone_url` checks to use the correct repo URL
```
Comment (JohanLorenzo) [inline procedures/automation-android-setup.md]: We may want to port the logic that we have in android-components to make this step seamless on other repos. Pointers: 

* https://github.com/mozilla-mobile/android-components/blob/bd045b5d820edeb05dfef3e254e2c5a0b5261d4a/.taskcluster.yml#L32-L34
* https://github.com/mozilla-mobile/android-components/blob/bd045b5d820edeb05dfef3e254e2c5a0b5261d4a/.taskcluster.yml#L102
* https://github.com/mozilla-mobile/android-components/blob/bd045b5d820edeb05dfef3e254e2c5a0b5261d4a/.taskcluster.yml#L111
* https://github.com/mozilla-mobile/android-components/blob/bd045b5d820edeb05dfef3e254e2c5a0b5261d4a/.taskcluster.yml#L232

File: procedures/automation-android-setup.md [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    2. Update scopes to be relevant for your product
+    3. Update the apks in `in.payload.command` so that the paths match your product
+    4. Update `in.payload.image` so that a compatible docker image is used
+    5. Update `in.metadata` for your product
+4. In `automation/taskcluster`, add `decision_task_nightly.py`, `schedule_nightly_graph.py` and `lib/tasks.py` that will perform most of the automation wiring. ([`fenix` examples](https://github.com/mozilla-mobile/fenix/blob/master/automation/taskcluster/))
```
Comment (JohanLorenzo) [inline procedures/automation-android-setup.md]: In the future we should avoid to create `decision_task_nightly.py`. Instead: 
* either we modify `decision.py` (example: https://github.com/mozilla-mobile/android-components/pull/1842) 
* or taskgraph will avoid the need of having decision_task_nightly.py

File: procedures/automation-android-setup.md [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+    1. `schedule_nightly_graph.py` is mostly good as-is, but I'd recommend a once-over to make sure there's nothing Fenix-specific
```
Comment (JohanLorenzo) [inline procedures/automation-android-setup.md]: @tomprince, out of curiosity, does https://hg.mozilla.org/ci/taskgraph include a way of scheduling cron tasks? If not, I don't think this is a high-priority. I'm just curious.

File: procedures/automation-android-setup.md [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+            2. Replace the second item in that list you pasted so that, instead of having the autograph username, it has `<%= scope.function_secret(["autograph_$product_prod_username"]) %>` (so it fetches from `hiera`)
+            3. Replace the third item in that list you pasted so that, instead of having the autograph password, it has `<%= scope.function_secret(["autograph_$product_prod_password"]) %>` (so it fetches from `hiera`)
+        5. Edit `manifests/settings.pp`, adding the new scope prefix `project:mobile:$product:releng:signing:` to the `scope_prefixes` property of both `mobile-dep` and `mobile-prod`
+
+    4. In `modules/pushapk_scriptworker`
+        1. From step 1, you should have received two certificates (one for dep, and one for prod). They start with `---BEGIN CERTIFICATE---` and end with `---END CERTIFICATE---`, and were probably sent in the gpg-encrypted text file with the autograph credentials. For each of these, copy them, remove any indentation they may have, and put them both in the `files` directory of `pushapk_scriptworker` with the names `$product_dep.pem` and `$product_release.pem`
```
Comment (JohanLorenzo) [inline procedures/automation-android-setup.md]: Nit: let's add a note to say that the certificate is public, even though it may have been sent encrypted.

File: procedures/automation-android-setup.md [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+            3. Add an entry to the `java_ks` block for your product, setting `certificate` to your certificate location
+    5. In `files/requirements.txt`
+        1. From step 9, update the version of `mozapkpublisher` and `pushapkscript`
+11. Commit and push your `build-puppet` changes, make a PR
+12. Test your automation with staging releases
+    1. Confirm with the team that nobody is using the dep signing or pushapk workers
+    2. SSH into the puppet master
+    3. Go to `/etc/puppet/environments/$username`
+    4. Temporarily move files out of the directory (like `environment.conf`)
+    5. `git clone https://github.com/$username/build-puppet .` to clone `build-puppet` to the directory
```
Comment (JohanLorenzo) [inline procedures/automation-android-setup.md]: Tip: You may want to add the upstream repo too. This helps in getting the latest changes when you want to work on a new bug.

File: procedures/automation-android-setup.md [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+    6. Checkout the branch you just pushed in step 11
+    7. Move the files back into the directory
+    8. Link config and nodes from the "main" environment
+        1. `cd manifests`
+        2. `ln -s moco-config.pp config.pp`
```
Comment (JohanLorenzo) [inline procedures/automation-android-setup.md]: The following notes are about the same as https://wiki.mozilla.org/ReleaseEngineering/PuppetAgain/HowTo/Set_up_a_user_environment. Maybe we just want to put a pointer here.

File: procedures/automation-android-setup.md [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+    1. Change the version code to be set to 1. If the product uses the same version-code-by-date schema as `fenix`, then edit [`versionCode.gradle`](https://github.com/mozilla-mobile/fenix/pull/156/files#diff-63606bb315fadc051f73a54767849985R41)
+    2. [Disable the creation of the task that pushes to Google Play](https://github.com/mozilla-mobile/fenix/pull/156/files#diff-73e51d972c105de5122ec559909980daR123)
+    3. Create the PR
+    4. Once approved, merge the PR
+16. Verify the apk artifact(s) of the signing task
+    1. Trigger the non-staging hook
+    2. Once the build finishes, download the apks from the signing task
+    3. Using the prod certificate from step 10.4.1., create a temporary keystore: `keytool -import -noprompt -keystore tmp_keystore -storepass 12345678 -file $product_release.pem -alias $product-rel`
```
Comment (JohanLorenzo) [inline procedures/automation-android-setup.md]: Nit: Github renders the numbered list as 10.iv.a (instead of 10.4.1). I'm unsure on how to properly fix this one. It depends on the configuration of the markdown parser

File: procedures/automation-android-setup.md [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+        * The "Digest algorithm" matches step 1
+        * The "Signature algorithm" matches step 1
+        * There are no warnings that there are entries "whose certificate chain invalid", "that are not signed by alias in this keystore" or "whose signer certificate is self-signed"
+    * Do the same thing for the dep signing task and certificate and check that the `jarsigner` command shows that the "Signed by" `CN` is "Throwaway Key"
+17. Request both the creation of a Google Play product and for the credentials to publish to it. Consult with the product team to [fill out the requirements for adding an app to Google Play](https://wiki.mozilla.org/Release_Management/Adding_a_new_app_on_Google_play). This request should be a bug for "Release Engineering > Release Automation: Pushapk", and should be a combination of [this](https://bugzilla.mozilla.org/show_bug.cgi?id=1508294) and [this](https://bugzilla.mozilla.org/show_bug.cgi?id=1512173)
+    * As part of the bug, note that you'll directly send an APK to Sylvestre Ledru and Julien Cristau via Slack
```
Comment (JohanLorenzo) [inline procedures/automation-android-setup.md]: Nit: We may want to replace "Sylvestre and Julien" by "the point of contact in release management". We don't know for sure whether Sylvestre and Julien will remain the point of contacts forever :slightly_smiling_face:

File: procedures/automation-android-setup.md [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+    4. `sudo cp /etc/hiera/secrets.eyaml /etc/hiera/secrets.eyaml.$username`, substituting your username in to back up the hiera secrets file
+    5. `sudo vi /etc/hiera/secrets.eyaml`, replace the dummy `service_account` and `certificate` values
+        * Reminder to properly indent these values to match other Google Play credentials in the file
+    6. `shred -u $p12file` to securely clean up the plaintext p12 key on the puppet master
+    7. `shred -u $p12file` wherever you decrypted it on your machine (you may need to install `shred`)
+20. Perform a new PR that un-stunts the changes from step 15 [Fenix example](https://github.com/mozilla-mobile/fenix/pull/161)
+    * Version code should be generated according to how the team requested in step 14
+    * The task that pushes to Google Play should no longer be disabled
+21. Once the PR from the last step is merged, trigger the non-staging nightly task, verify that it uploads to Google Play
+22. Update the `$product-nightly` hook, adding a schedule of `0 12 * * *` (make it fire every night)
```
Comment (JohanLorenzo) [inline procedures/automation-android-setup.md]: Nit: s/every night/every day at noon UTC

File: procedures/automation-android-setup.md [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+
+Throughout this document, wherever the term `$product` is used, substitute your product's name in (replacing spaces with hyphens), e.g.: `reference-browser` or `fenix`.
+
+Note: we don't need to explicitly define scopes in Taskcluster. We just define them in the task, hook's role, and each scriptworker script. Taskcluster will implicitly detect scopes, but it doesn't use them for anything other than generic equality/wildcard verification.
+
+1. Request dep and release signing keys (example bug for [`reference-browser`](https://bugzilla.mozilla.org/show_bug.cgi?id=1508761)). Confirm with the app's team what the "signature algorith" and "digest algorithm" should be, and include that information in the ticket.
+2. Clone the product's repository
+    * Note that, while doing "staging" runs, you'll have to use your `master` branch in your fork. Hook cloning/CoT/etc haven't been tested with a non-default branch
```
[the remark this answers is not in the document]
Comment (mitchhentges) [reply procedures/automation-android-setup.md]: Hmm, I'm not sure how the config and code I point to in the rest of the guide work with that - I might be handling a clone/commit the wrong way, which could :b:ork it.
I'll change this accordingly to just recommend using the `master` branch.
Maybe after taskgraph-ing, the in-repository logic will be simple enough that I can more-confidently recommend using a non-master branch!

[the remark this answers is not in the document]
Comment (JohanLorenzo) [reply procedures/automation-android-setup.md]: Ah yeah, I confirm this doesn't work as is on Fenix. I just encountered this error https://tools.taskcluster.net/groups/c_FN0u-WTBuqQoOryKBXnQ/tasks/c_FN0u-WTBuqQoOryKBXnQ/runs/0/logs/public%2Flogs%2Flive_backing.log#L437 because .taskcluster.yml re-clones the forked repo. I'll see if I can make this better. I need it to test https://github.com/mozilla-mobile/fenix/pull/397 out

[the remark this answers is not in the document]
Comment (JohanLorenzo) [reply procedures/automation-android-setup.md]: Great news! The effort to allow branches to work is minimal: https://github.com/mozilla-mobile/fenix/pull/397/commits/55b419cfcab926ddc73add838db05be65da1e3a1

Bonus points: we can reuse the repo baked in the docker image. This saves some network and CPU cycles \o/

File: procedures/automation-android-setup.md [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+        1. Make PRs: [example `pushapkscript` release PR](https://github.com/mozilla-releng/pushapkscript/pull/70)
+        2. Once these release PRs are CR'd and merged, publish each package:
+            1. `git checkout master`, make sure you're on the "Merge pull request" commit for your PR
+            2. `git tag $version`, e.g.: `git tag 0.11.0`
+            3. `git push --tags && git push upstream --tags` (assuming that the `origin` remote is for your fork, and `upstream` is the `mozilla-releng` repo)
+            4. Ensure you're in the Python virtual env for your package (One approach is to share a single virtual env between all scriptworker repos)
+            5. `rm -rf dist && python setup.py sdist bdist_wheel` build the package
+            6. Publish to `Pypi`:
+                1. `gpg --list-secret-keys --keyid-format long` to get your GPG identity (it's the bit after "sec   rsaxxxx/"). An example GPG identity would be `5F2B4756ED873D23`
+                2. `twine upload --sign --identity $identity dist/*` to upload to Pypi (you may need to `pip install twine` first)
```
Comment (MihaiTabara) [inline procedures/automation-android-setup.md]: ++ for adding these steps.

File: procedures/automation-android-setup.md [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+        3. Update the scope to assume from the non-staging hook's role
+7. Create the role for the staging hook (e.g.: [`fenix-nightly-staging`](https://taskcluster-web.netlify.com/auth/roles/hook-id%3Aproject-mobile%2Ffenix-nightly-staging))
+    1. Go to [Roles](https://taskcluster-web.netlify.com/auth/roles)
+    2. `RoleId` should be idential to your staging hook's `HookId`, but prefixed with `hook-id:` (e.g.: `hook-id:project-mobile/fenix-nightly-staging`)
+    3. `Description` should be something like "Hook for building, signing and uploading (staging) Nightly versions of $product"
+    4. Add all the scopes you'll need to fulfill the requirements in you `.taskcluster.yml` file (when `is_mozilla_mobile_repo` is `false`)
+8. Do the same thing, but for non-staging nightlies (e.g.: [`fenix-nightly`](https://taskcluster-web.netlify.com/auth/roles/hook-id%3Aproject-mobile%2Ffenix-nightly))
+    1. Remove mentions of "staging" from each field
+    2. Make sure that all scopes are provided to fulfill requirements in `.taskcluster.yml` (when `is_mozilla_mobile_repo` is `true`)
+9. Update `scriptworker`, `pushapk` and `mozapkpublisher`
```
Comment (MihaiTabara) [inline procedures/automation-android-setup.md]: I'm wondering if we should add a quick note like: _Should you need publishing an Android product to the Google Play store, you'l also need to ..._ Mainly I'm thinking at `android-components` or `application-services` which don't have this step but instead they require `beetmover` bits. However, I could be wrong as this docs overall is aimed for the former.

File: procedures/automation-android-setup.md [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+            2. `git tag $version`, e.g.: `git tag 0.11.0`
+            3. `git push --tags && git push upstream --tags` (assuming that the `origin` remote is for your fork, and `upstream` is the `mozilla-releng` repo)
+            4. Ensure you're in the Python virtual env for your package (One approach is to share a single virtual env between all scriptworker repos)
+            5. `rm -rf dist && python setup.py sdist bdist_wheel` build the package
+            6. Publish to `Pypi`:
+                1. `gpg --list-secret-keys --keyid-format long` to get your GPG identity (it's the bit after "sec   rsaxxxx/"). An example GPG identity would be `5F2B4756ED873D23`
+                2. `twine upload --sign --identity $identity dist/*` to upload to Pypi (you may need to `pip install twine` first)
+
+10. Update configuration in [`build-puppet`](https://github.com/mozilla-releng/build-puppet/)
+    1. Locate signing secrets (there's six different items you'll add to `hiera`)
```
Comment (MihaiTabara) [inline procedures/automation-android-setup.md]: Nit: should we expand for clarity what those six items are? Even though you refer to them immediately below, it might be useful to clarify this even more.

File: procedures/automation-android-setup.md [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+        1. You should've received signing credentials from step 1. Print out the decrypted file you received: `gpg -d <file from step 1>`
+        2. We will want to encrypt the "dep" and "rel" credentials for the "prod" autograph instance. They can be identified as lines that contain a "list" where the second item ends with "_dep" or "_prod", respectively
+            * Example: "dep" line would be: `["http://<snip>", "signingscript_fenix_dep", "<snip>", ["autograph_apk"], "autograph"]`
+        3. For these two lines, the secrets we want to put in hiera are the username and password (the second and third item)
+        4. Later, in step 19, you'll have been emailed a Google Play service account and key. However, for now, we're going to use a dummy value (the string "dummy") as placeholders for these values
+
+    2. Add secrets to `hiera`
+        1. Connect to VPN
+        2. SSH into `releng-puppet2.srv.releng.mdc1.mozilla.com`
+        3. `sudo cp /etc/hiera/secrets.eyaml /etc/hiera/secrets.eyaml.$username` (substituting your username in) to back up the hiera secrets file
```
Comment (MihaiTabara) [inline procedures/automation-android-setup.md]: Nit: Might be useful to have some timeline baked there as well, something like `/etc/hiera/secrets.eyaml.$username.bak-DD-MM-YYY` or alike.

File: procedures/automation-android-setup.md [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+        4. For each secret, encrypt it with `sudo eyaml encrypt --pkcs7-private-key /etc/hiera/keys/private_key.pem --pkcs7-public-key /etc/hiera/keys/public_key.pem --output examples -p -l '$lookupkey'` (it prompts you to paste the secret). The "lookup key" is different for each secret we put in Hiera:
```
Comment (MihaiTabara) [inline procedures/automation-android-setup.md]: Nit: Might be worth to also pointing to [this](https://wiki.mozilla.org/ReleaseEngineering/PuppetAgain/Secrets#Using_EYAML) for more docs on this matter?

File: procedures/automation-android-setup.md (hunk 6 above)
Comment (MihaiTabara) [reply procedures/automation-android-setup.md]: +1

File: procedures/automation-android-setup.md [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+    6. Checkout the branch you just pushed in step 11
+    7. Move the files back into the directory
+    8. Link config and nodes from the "main" environment
+        1. `cd manifests`
+        2. `ln -s moco-config.pp config.pp`
+        3. `ln -s moco-nodes.pp nodes.pp`
+    9. Edit `manifests/moco-nodes.pp` to pin both `dep-m-signing-linux` and `dep-m-pushapkworker` to your environment
+        * Add `$pin_puppet_server = 'releng-puppet2.srv.releng.mdc1.mozilla.com'`
+        * Add `$pin_puppet_env = '$username'`
+    10. SSH into both the pushapk worker (`dep-m-pushapkworker-1.srv.releng.use1.mozilla.com`) and the signing worker (`dep-m-signing-linux-1.srv.releng.use1.mozilla.com`)
```
Comment (MihaiTabara) [inline procedures/automation-android-setup.md]: Tip: if we later add more instances and ssh-ing to each of them becomes impracticle, this can also be achieved by using `csshX`.

```
# vpn
for i in {1..10}; do
nslookup dep-m-pushapkworker-$i | grep Name: | sed -e 's/Name:\t//'
done > /tmp/dep-pushapkworker
csshX --hosts /tmp/dep-pushapkworker
sudo puppet agent --test  # or unpin or w/e
```

File: procedures/automation-android-setup.md [hunk 17]
```diff
@@ -0,0 +1,200 @@
...
+    1. Change the version code to be set to 1. If the product uses the same version-code-by-date schema as `fenix`, then edit [`versionCode.gradle`](https://github.com/mozilla-mobile/fenix/pull/156/files#diff-63606bb315fadc051f73a54767849985R41)
+    2. [Disable the creation of the task that pushes to Google Play](https://github.com/mozilla-mobile/fenix/pull/156/files#diff-73e51d972c105de5122ec559909980daR123)
+    3. Create the PR
+    4. Once approved, merge the PR
+16. Verify the apk artifact(s) of the signing task
+    1. Trigger the non-staging hook
+    2. Once the build finishes, download the apks from the signing task
+    3. Using the prod certificate from step 10.4.1., create a temporary keystore: `keytool -import -noprompt -keystore tmp_keystore -storepass 12345678 -file $product_release.pem -alias $product-rel`
+    4. For each apk, verify that it matches the certificate: `jarsigner -verify $apk -verbose -strict -keystore tmp_keystore`. Check that
```
Comment (MihaiTabara) [inline procedures/automation-android-setup.md]: ++ for adding these steps

File: procedures/automation-android-setup.md [hunk 18]
```diff
@@ -0,0 +1,200 @@
...
+    * As part of the bug, note that you'll directly send an APK to Sylvestre Ledru and Julien Cristau via Slack
+18. Give the first signed APK to the Google Play admins
+    1. Perform a non-staging build
+    2. Once the signing task is done, grab the APK with the version code of 1 (if there's multiple APKs, you probably want the arm one)
+        * You can verify the version code of the apk with [apktool](https://ibotpeaches.github.io/Apktool/), then viewing the extracted `AndroidManifest.xml` and looking at the `platformBuildVersionCode`
+    3. Send the APK to Sylvestre and Julien
+19. Once the previous step is done and they've set up a Google Play product, put the associated secrets in Hiera
+    1. Connect to VPN and SSH into the puppet master
+    2. Encrypt the `service_account` (you'll have been emailed or slack'd a google service account: it looks like an email address that ends in `gserviceaccount.com`)
+        * `sudo eyaml encrypt --pkcs7-private-key /etc/hiera/keys/private_key.pem --pkcs7-public-key /etc/hiera/keys/public_key.pem --output examples -p -l 'service_account'`
```
Comment (MihaiTabara) [inline procedures/automation-android-setup.md]: Nit: Maybe worth adding a pointer to the [official docs](https://wiki.mozilla.org/ReleaseEngineering/PuppetAgain/Secrets#Using_EYAML) on this for more tips?

File: procedures/automation-android-setup.md (hunk 1 above)
Comment (mitchhentges) [reply procedures/automation-android-setup.md]: I don't quite follow: 

* Are you referring to making `$repository` work for both hook and pull request triggers? In this case, since the documentation is specifically about setting up nightly, I'd like to opt towards leaving it simple and let a follow-up guide explain how to set up PR checks :)
* Alternatively, ado you mean how we need to have checks that the correct repo URL is used, and how we're directly referring to the repository (`https://github.com/mozilla-mobile/$project`)? I'm not sure that we can work around that, and `android-components` [has the same thing](https://github.com/mozilla-mobile/android-components/blob/bd045b5d820edeb05dfef3e254e2c5a0b5261d4a/.taskcluster.yml#L44) to check if the repo is trusted.

TL;DR I'm a nerd and I don't understand the recommendation :)

File: procedures/automation-android-setup.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (mitchhentges) [reply procedures/automation-android-setup.md]: Wow, I didn't know that you wouldn't have to re-clone like that! That's awesome! :tada: :tada:

File: procedures/automation-android-setup.md (hunk 2 above)
Comment (mitchhentges) [reply procedures/automation-android-setup.md]: I'd like to wait for `taskgraph` for this migration, if y'alls are :ok_hand: with that.

File: procedures/automation-android-setup.md (hunk 3 above)
Comment (mitchhentges) [reply procedures/automation-android-setup.md]: Arguably, that's not `taskgraph`'s responsibility, right? We have TC hooks which run the tasks according to cron. `schedule_nightly_graph` just invokes `.taskcluster.yml` with a bunch of Github environment variables set up

File: procedures/automation-android-setup.md (hunk 5 above)
Comment (mitchhentges) [reply procedures/automation-android-setup.md]: Hmm, I didn't realize that this section was on the puppet master. You have `upstream` configured there? I keep my local `build-puppet` up to date and strictly pull on my repo on `build-puppet` so I don't have conflicts.

Both approaches are valid, but since we aren't consistent here, I'll leave it up to the reader :)

File: procedures/automation-android-setup.md (hunk 7 above)
Comment (mitchhentges) [reply procedures/automation-android-setup.md]: I changed the text to `10.iv.a` ez :smile:

File: procedures/automation-android-setup.md [hunk 19]
```diff
@@ -0,0 +1,200 @@
+# How to set up automation for nightly Android product
+
+These instructions define which resources are needed for setting up nightly Android product release automation, as well as the steps to prepare the automation itself. Additionally, this guide include steps for running a "staging" version of the automation so that it can be tested in the developer's own repository.
```
[the remark this answers is not in the document]
Comment (mitchhentges) [reply procedures/automation-android-setup.md]: Sure, I'll emphasize that this is specific to Google Play nightly automation (plus staging releases)

File: procedures/automation-android-setup.md (hunk 12 above)
Comment (mitchhentges) [reply procedures/automation-android-setup.md]: Call me selfish, but I'd like to wait until we have our documentation situation sorted out. Once that's accomplished, we can restructure these docs to branch and be more flexible.

I'm guessing that, long-term, we'll have these split into sections like:
* Core automation
* PR checks
* Signing
* Deployment to Google Play
* Deployment to Amazon Store
* Deployment to Maven
* etc

What do you think?

File: procedures/automation-android-setup.md (hunk 14 above)
Comment (mitchhentges) [reply procedures/automation-android-setup.md]: I personally just pop it in `secrets.eyaml.mhentges`, and it looks like others do, too.
However, I can recommend using a date in these instructions

File: procedures/automation-android-setup.md (hunk 16 above)
Comment (mitchhentges) [reply procedures/automation-android-setup.md]: Since these are dep workers, I'm feeling lucky that we'll sit with one worker (at least for a while).
This is worth revisiting if we start adding additional dep workers

File: procedures/automation-android-setup.md (hunk 15 above)
Comment (mitchhentges) [reply procedures/automation-android-setup.md]: I'll link to the docs, but I'll still inline the commands so that it's easy to see what needs to be accomplished

File: procedures/automation-android-setup.md (hunk 18 above)
Comment (mitchhentges) [reply procedures/automation-android-setup.md]: I'll add a link to the encryption docs here, too, good call

File: procedures/automation-android-setup.md (hunk 3 above)
Comment (tomprince) [reply procedures/automation-android-setup.md]: taskgraph will have the same support for scheduling cron graphs (in conjunction with ci-admin):
- ci-admin creates a hook that has job that runs `taskgraph cron`
- `taskgraph cron` looks at `.cron.yml` in the appropriate repository and creates any jobs that are schedule then, using `.taskcluster.yml` for the task description.

It sounds like `schedule_nightly_graph` is analogous to `taskgraph cron` here.

Comment (mitchhentges) [reply procedures/automation-android-setup.md]: Sounds like I'm a little out-of-date here: I'm familiar with cron timings being set in the hook (which I suppose will be ci-admin'd). Once the job actually starts, it's just a regular hook trigger.

What kind of configuration is stored in `.cron.yml`? Does it change what jobs will run based on when a hook is triggered?

File: procedures/automation-android-setup.md (hunk 2 above)
Comment (MihaiTabara) [reply procedures/automation-android-setup.md]: Sounds good 👍

File: procedures/automation-android-setup.md (hunk 3 above)
Comment (MihaiTabara) [reply procedures/automation-android-setup.md]: @mitchhentges https://hg.mozilla.org/mozilla-central/file/tip/.cron.yml might be useful to glance at in this context. I'd assume it's going to be something similar.

File: procedures/automation-android-setup.md (hunk 12 above)
Comment (MihaiTabara) [reply procedures/automation-android-setup.md]: This is great, indeed! I fully agree 👍

File: procedures/automation-android-setup.md [hunk 20]
```diff
@@ -0,0 +1,200 @@
...
+
+Throughout this document, wherever the term `$product` is used, substitute your product's name in (replacing spaces with hyphens), e.g.: `reference-browser` or `fenix`.
+
+Note: we don't need to explicitly define scopes in Taskcluster. We just define them in the task, hook's role, and each scriptworker script. Taskcluster will implicitly detect scopes, but it doesn't use them for anything other than generic equality/wildcard verification.
```
[the remark this answers is not in the document]
Comment (MihaiTabara) [reply procedures/automation-android-setup.md]: Ah, ok, this makes sense, sorry about the misunderstanding. I still think it's worth adding a small sentence talking about the manuevring of the scopes and the control enforcing, not _defining_ them. But we can always iterate later, this is certainly not blocking.

[the remark this answers is not in the document]
Comment (mitchhentges) [reply procedures/automation-android-setup.md]: I don't quite understand: what kind of thing would you like me to emphasize? I'm not sure what to place about "manuevering the scopes and control enforcing" - is this about ci-admin? :thinking:

File: procedures/automation-android-setup.md (hunk 16 above)
Comment (MihaiTabara) [reply procedures/automation-android-setup.md]: Makes total sense, I added it more for the future context 👍

File: procedures/automation-android-setup.md (hunk 3 above)
Comment (mitchhentges) [reply procedures/automation-android-setup.md]: Wait, so we'll have _some_ cron-related build configuration in-repository, and some of it on our hooks (so, in ci-admin)? Doesn't this fragment our config a bit?

Comment (tomprince) [reply procedures/automation-android-setup.md]: For mozilla-central, the hooks are run every 15m, and then the task that runs is responsible for figuring out if there is anything scheduled at that time.

Comment (mitchhentges) [reply procedures/automation-android-setup.md]: Ohhh, interesting - would it be easier to have a separate hook for each build, so we don't need cron configuration in our repositories?

Comment (tomprince) [reply procedures/automation-android-setup.md]: I would be inclined to say it is easier the other way around: there is a standard hook that doesn't have any project specific details, and then everything can be configured from the repository, rather than needing to change hooks to get changed behavior.

Comment (mitchhentges) [reply procedures/automation-android-setup.md]: I feel like this solution is causing us to stack two solutions on top of each other where a single standalone one would work better.

For example, some downsides we're going to see are:
* artificial delays before builds start. We need to wait for the next 15 minute trigger from the _outside_ cron system before our _internal_ cron system can trigger builds (that potentially could've happened 14 minutes ago). Tweaking an arbitrary "poll" value isn't ideal
* We lose the benefits of the outside hook system:
    * I can't just "trigger" a hook if I need to run it early
    * I can't click "view last run" of a certain hook to see what went wrong
* We need to maintain our internal cron system
* Explaining the system to new relengers/devs: "You customize your cron job here in `.cron.yml`. No, our internal cron system doesn't actually start the build. Yeah, an entirely separate system that _also_ runs on a timer triggers our internal system, which delegates accordingly"
* FWIW, though minimal: when our 15m hook triggers, but no jobs need to run, it just uses unnecessary machine hours and populates our job-tracking systems with empty builds

Although, I can definitely see the benefit of having a `.cron.yml` in your repository: it's much better for reviewing changes and tracking history. Would our time making an internal cron system be better spent improving Taskcluster integration (e.g.: list each cron job separately as its own "virtual" hook?)

File: procedures/automation-android-setup.md (hunk 20 above)
[the remark this answers is not in the document]
Comment (MihaiTabara) [reply procedures/automation-android-setup.md]: Yes. I see this section slightly expanded:

a) existing paragraph that describes what scopes actually are and how they don't need to be defined specifically, but are strings in Taskcluster that we define in various places to ensure a good security model.

b) a follow-up paragraph like the one I mentioned above, where we complete moreover the fact that these scope administration in hooks/roles and such is done or will be done soon via tools like `ciadmin`.

[the remark this answers is not in the document]
Comment (mitchhentges) [reply procedures/automation-android-setup.md]: I'll add a note, though there are other changes that will affect these docs as well, such as the `pushapk` changes - rather than having a note of what _will_ change, I'd rather update the docs _when_ the change happens. This way, the docs stay tight and relevant

[the remark this answers is not in the document]
Comment (MihaiTabara) [reply procedures/automation-android-setup.md]: That's a fair point. I guess I'm being biased since I hope to land those changes this week :)

File: procedures/automation-android-setup.md (hunk 1 above)
Comment (JohanLorenzo) [reply procedures/automation-android-setup.md]: I'm sorry. Even I was confused when I read my comment. I mean: if we extract the repository from the Github event and we pass it down to `decision.py`, there is no need to ``event.repository.html_url` at all. It just works without having to temporary modify some params.

Comment (mitchhentges) [reply procedures/automation-android-setup.md]: Sounds good: If you're on-board, I want the docs to point to "Fenix" as the reference for automation setup, because:
1. I'm biased, I set it up :stuck_out_tongue: 
2. `android-components` has more mature automation, but it has the complexity of a variable number of sub-projects (and it doesn't do pushapk). I'd rather that our reference be more like a standalone product
3. `reference-browser` is less production-y than Fenix

When we update Fenix's `decision.py` and how it works, let's update the docs, too!

**Comment by mitchhentges** [comment]
I'm not going to merge this until [RFC 7](https://github.com/mozilla-releng/releng-rfcs/pull/7) is resolved

**Comment by mitchhentges** [comment]
Discussion in [RFC 7](https://github.com/mozilla-releng/releng-rfcs/pull/7) decided that most docs would belong to `build-relengdocs`, and my interpretation is that these docs are included.

@JohanLorenzo I'd like to land these as-is (very obsolete), then do a follow-up PR to modernize it. That way:
1. If something more important comes up, at least we have _some_ docs live
2. We have history in the repository of how automation used to tick, which _could_ be useful?

**Comment by JohanLorenzo** [comment]
👍 This plan makes sense to me
